changefeedccl: support a CSV format for changefeeds

sherman-grewal opened the pull request:
In this PR, we introduce a new CSV format for changefeeds.
Note that this format is only supported with the
initial_scan='only' option. For instance, one can now
execute:

CREATE CHANGEFEED FOR foo WITH format=csv, initial_scan='only';

Release note (enterprise change): Support a CSV format for
changefeeds, with the intention of supporting use cases 
similar to EXPORT. Only works with initial_scan='only', 
and does not work with diff/resolved options.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79853)
<!-- Reviewable:end -->

miretskiy commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -820,6 +820,22 @@ func validateDetails(details jobspb.ChangefeedDetails) (jobspb.ChangefeedDetails
+				return jobspb.ChangefeedDetails{}, err
+			}
+			if initialScanType != changefeedbase.OnlyInitialScan {
+				return jobspb.ChangefeedDetails{}, errors.Errorf(
```
I wonder why do you feel we have to have this restriction? is this purely because .. we don't know how to handle schema changes?

however cloud storage sinks name files with the "schema id" -- so changes in schema can be handled.

sherman-grewal replied:
Yeah this might be possible to address with cloud storage sinks, but what about non cloud storage sinks? When we address this issue for all of the sinks I would be more inclined to drop this restriction.

miretskiy commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -820,6 +820,22 @@ func validateDetails(details jobspb.ChangefeedDetails) (jobspb.ChangefeedDetails
+					`%s=%s is only usable with %s='only'`,
+					changefeedbase.OptFormat, changefeedbase.OptFormatCSV, changefeedbase.OptInitialScan)
+			}
+			if _, ok := details.Opts[changefeedbase.OptResolvedTimestamps]; ok {
```
Why can't we support this?

miretskiy commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -571,18 +578,30 @@ func (s *webhookSink) workerLoop(workerIndex int) {
+				encoded.alloc.Release(s.workerCtx)
+				s.metrics.recordEmittedBatch(
+					encoded.emitTime, len(msgs), encoded.mvcc, len(encoded.data), sinkDoesNotCompress)
+			case changefeedbase.OptFormatCSV:
```
can we change encodePayloadWebhook to handle both formats? This way, we won't have to duplicate sendmessagewith retries.. and also, forget to record metrics?

Alternatively:
```
  batchEncoder := func() func(messages []messagePayload) (encodedPayload, error)  {
      if JSON {
         return encodePayloadJson
     }
     return encodePayloadCSV
}()
```
Then just use batchEncoder() instead of calling encodePayloadWebhook.

sherman-grewal replied:
Thanks for the suggestion! I completely agree 🙌🏽

HonoreDB reviewed (changes requested):
Copy/pasting from slack for posterity: My thinking here is that resolved and diff don't make any sense as options for initial_scan=only anyway--we won't emit resolved timestamps and the before in initial_scan mode is meaningless. So disallow them for initial_scan=only, which in turn means we don't have to figure out how to encode them in csv right now. Then later if we need to support those options for csv, we can figure out an encoding scheme.

> Reviewed 2 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-:-N-UID-g4cl53N-gM9Gm:b2ufqat)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sherman-grewal and @stevendanna)

___
*[pkg/ccl/changefeedccl/encoder.go, line 373 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UAJ__76dGXTbgMD6s:-N-UAJ__76dGXTbgMD6t:bv6u1xu) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/encoder.go#L373)):*
> ```Go
> }
> 
> type csvEncoder struct {
> ```

Format number 3 (ish) strikes me as the right moment to break out different encoders into their own files.
___
*[pkg/ccl/changefeedccl/encoder.go, line 409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UFdlY-ajU-turPcLG:-N-UFdlY-ajU-turPcLH:bnrawo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/encoder.go#L409)):*
> ```Go
> 	include := make(map[descpb.ColumnID]struct{}, len(family.ColumnIDs))
> 	var yes struct{}
> 	for _, colID := range family.ColumnIDs {
> ```

Could take advantage of the fact that this is restricted to `initial_scan=only` to cache this map--we shouldn't be seeing multiple versions of a table. (There might be 2? Not 3.)
___
*[pkg/ccl/changefeedccl/encoder.go, line 413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UC-OQ9wXNtVx4mNMw:-N-UC-OR2E0vo-QmfVce:b-idw1uc) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/encoder.go#L413)):*
> ```Go
> 	}
> 
> 	var csvRow []string
> ```

Is there a way to do this without building this slice? Does the csv writer support something like `writer.WriteValue(tree.AsString(datum.Datum))` and then `writer.EndOfRow()` after we're done?
___
*[pkg/ccl/changefeedccl/encoder.go, line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UGDJw6FOKe3K_2nh-:-N-UGDJw6FOKe3K_2nh0:b4b8ha1) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/encoder.go#L439)):*
> ```Go
> 	_ context.Context, _ string, resolved hlc.Timestamp,
> ) ([]byte, error) {
> 	return nil, nil
> ```

Probably better to error here.
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 5868 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UGzaH2veSe6MxCIc1:-N-UGzaH2veSe6MxCIc2:bqgz4np) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/changefeed_test.go#L5868)):*
> ```Go
> 				sqlDB.Exec(t, "INSERT INTO foo VALUES (4, 'Doug'), (5, 'Elaine'), (6, 'Fred')")
> 
> 				expectedMessages := []string{
> ```

So we're not expected to emit a header message, I take it? It'd be a little tricky but we could.


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/changefeedccl/encoder.go:
```diff
@@ -367,6 +370,75 @@ func (e *jsonEncoder) getTableColMap(desc catalog.TableDescriptor) catalog.Table
+
+// EncodeValue implements the Encoder interface.
+func (e *csvEncoder) EncodeValue(_ context.Context, row encodeRow) ([]byte, error) {
+	if row.deleted {
```
I guess that's the reason why we only support initial scan.
it's a pity, but I think it's reasonable for now.

miretskiy reviewed (commented):
This is very nice.  Few nits/questions...

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79853)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @stevendanna)

___
*[pkg/ccl/changefeedccl/encoder.go, line 373 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UAJ__76dGXTbgMD6s:-N-ck8kE5aW298HT91_5:bkqxko) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/encoder.go#L373)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Format number 3 (ish) strikes me as the right moment to break out different encoders into their own files.
</blockquote></details>

yeah I totally agree - I added a new files for the three encoders. Thanks for the suggestion :smile: 
___
*[pkg/ccl/changefeedccl/encoder.go, line 409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UFdlY-ajU-turPcLG:-N-ckGCkBkgR3-objSwR:b-frteys) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/encoder.go#L409)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Could take advantage of the fact that this is restricted to `initial_scan=only` to cache this map--we shouldn't be seeing multiple versions of a table. (There might be 2? Not 3.)
</blockquote></details>

Yeah I added a mechanism to cache the map (which is then rebuilt if we see a new version number). If we see another version, does it mean that we won't see rows with the previous version number? If we can still see both version numbers, I was considering storing the cache for each version we see (which is at most 2). Let me know what you think.
___
*[pkg/ccl/changefeedccl/encoder.go, line 413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UC-OQ9wXNtVx4mNMw:-N-ckdWT7-cUKvt43yQI:bm3l45v) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/encoder.go#L413)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Is there a way to do this without building this slice? Does the csv writer support something like `writer.WriteValue(tree.AsString(datum.Datum))` and then `writer.EndOfRow()` after we're done?
</blockquote></details>

As mentioned on slack, there is no way of building this without a slice if we want to use the CSV writer package. I added a TODO to keep a lookout of the performance of CSV encoding for large rows - if we incur a significant loss in performance we can look into building our own CSV writer that writes values directly into the buffer rather than storing them in an array and then writing the entire array to the buffer.
___
*[pkg/ccl/changefeedccl/encoder.go, line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UGDJw6FOKe3K_2nh-:-N-cl02T0DlzuTz14WXR:bcivurw) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/encoder.go#L439)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Probably better to error here.
</blockquote></details>

Done - Thanks for the suggestion!
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 5868 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UGzaH2veSe6MxCIc1:-N-cl55UANmUUdJ7r0mX:b-sgsn9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/changefeed_test.go#L5868)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

So we're not expected to emit a header message, I take it? It'd be a little tricky but we could.
</blockquote></details>

As mentioned on slack, we decided to not emit a header for a few reasons
- The EXPORT statement does not do this
- We need to establish the product requirements for the behaviour of header messages in file-based sinks (e.g. do we re-emit the header message for each new file we create? or is it done once?)
- We need to establish the behaviour of re-emitting the header message if the changefeed was paused and then unpaused

I think that once we iron these issues out we can provide an option to include a header message.


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (commented):
> Reviewed 3 of 11 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79853)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @stevendanna)

___
*[pkg/ccl/changefeedccl/encoder.go, line 409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UFdlY-ajU-turPcLG:-N-cmGerC774SpRrLovi:botdtni) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/encoder.go#L409)):*
<details><summary><i>Previously, sherman-grewal (Sherman) wrote…</i></summary><blockquote>

Yeah I added a mechanism to cache the map (which is then rebuilt if we see a new version number). If we see another version, does it mean that we won't see rows with the previous version number? If we can still see both version numbers, I was considering storing the cache for each version we see (which is at most 2). Let me know what you think.
</blockquote></details>

I'd just store the cache for each version. If it's possible to see multiple versions it's possible to see them out of order.
___
*[pkg/ccl/changefeedccl/encoder_avro.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-cn10U1b8V5Ll8iJl_:-N-cn10U1b8V5Ll8iJla:bwvh7od) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee566a19ccbb2c079e25b2adf7269ca4ded874a1/pkg/ccl/changefeedccl/encoder_avro.go#L131)):*
> ```Go
> // Get the raw SQL-formatted string for a table name
> // and apply full_table_name and avro_schema_prefix options
> func (e *confluentAvroEncoder) rawTableName(
> ```

Oh man did I really not delete this function when I was doing the topic refactor? Whoops. This should just be a call to TopicNamer.Name but I can do that in a separate PR.


<!-- Sent from Reviewable.io -->

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79853)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @stevendanna)

___
*[pkg/ccl/changefeedccl/encoder.go, line 409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-UFdlY-ajU-turPcLG:-N-cxoXB77DQ5tjxRKRg:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/87326bb17c7bd3c2ff8a5a1f53af2ae458d96460/pkg/ccl/changefeedccl/encoder.go#L409)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I'd just store the cache for each version. If it's possible to see multiple versions it's possible to see them out of order.
</blockquote></details>

Done
___
*[pkg/ccl/changefeedccl/encoder_avro.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-cn10U1b8V5Ll8iJl_:-N-cxjt24X8lIaD7pfbE:b-fuadhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee566a19ccbb2c079e25b2adf7269ca4ded874a1/pkg/ccl/changefeedccl/encoder_avro.go#L131)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Oh man did I really not delete this function when I was doing the topic refactor? Whoops. This should just be a call to TopicNamer.Name but I can do that in a separate PR.
</blockquote></details>

ahh ok sounds good, thanks for catching that!!


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -124,6 +126,21 @@ func encodePayloadWebhook(messages []messagePayload) (encodedPayload, error) {
+	var mergedMsgs []byte
+	for _, m := range messages {
+		result.alloc.Merge(&m.alloc)
+		mergedMsgs = append(mergedMsgs, m.val...)
```
looks like we are not updating some of the fields in the result -- such as mvcc timestamp.

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79853)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @stevendanna)

___
*[pkg/ccl/changefeedccl/encoder_csv.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-Z_FvCVxB47GG5oydI-r3-91:-N-dv39v3txggVcjM30I:bndt16u) ([raw file](https://github.com/cockroachdb/cockroach/blob/74881abf4e54d7594ad37e8f1b749f4db7dcdab9/pkg/ccl/changefeedccl/encoder_csv.go#L91)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

At the very least, we should memoize this array; store csvRow array in the encoder struct, initialize it with `csvRow: make([]string, 0, len(row.tableDesc.PublicColumns()))` and then zero it out here:
```
csvRow = csvRow[:0]
```

</blockquote></details>

I completely agree, I implemented memoization in this array in my latest commit. Thanks again for the suggestions :+1: 
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 137 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N-U3Pv9FVh5SZt5s8jO-r3-137:-N-duw0ADAZgwPZEqE3s:bmrhg85) ([raw file](https://github.com/cockroachdb/cockroach/blob/74881abf4e54d7594ad37e8f1b749f4db7dcdab9/pkg/ccl/changefeedccl/sink_webhook.go#L137)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

looks like we are not updating some of the fields in the result -- such as mvcc timestamp.  
</blockquote></details>

ahh yes - thanks for catching this! :raised_hands: 


<!-- Sent from Reviewable.io -->

miretskiy commented:
@sherman-grewal are you waiting for additional comments?

HonoreDB reviewed (approved):
LGTM after one comment is addressed, I'll pre-emptively approve it too.

> Reviewed 2 of 9 files at r1, 6 of 11 files at r2, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-:-N0aQV_l3B5HQ_SDbdYn:b-10kr34)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @stevendanna)

___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 5870 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-N0aQ8YNBleuG9ujBKTd:-N0aQ8YNBleuG9ujBKTe:b-lp5o5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/45e5c77d227c16d98836b8a4e80f417d2fcd5585/pkg/ccl/changefeedccl/changefeed_test.go#L5870)):*
> ```Go
> 		`initial scan only with csv`:     `CREATE CHANGEFEED FOR foo WITH initial_scan_only, format = csv`,
> 		`initial backfill only with csv`: `CREATE CHANGEFEED FOR foo WITH initial_scan = 'only', format = csv`,
> 	}
> ```

Mind also adding a test with one changefeed on two different tables? It'll be one encoder for both tables so there can be bugs there.


<!-- Sent from Reviewable.io -->

Commit:
changefeedccl: support a CSV format for changefeeds
In this PR, we introduce a new CSV format for changefeeds.
Note that this format is only supported with the
initial_scan='only' option. For instance, one can now
execute:

CREATE CHANGEFEED FOR foo WITH format=csv, initial_scan='only'

Release note (enterprise change): Support a CSV format for
changefeeds. Only works with initial_scan='only', and
does not work with diff/resolved options.

sherman-grewal reviewed (commented):
Done, thanks for the reviews!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79853#-:-N0aX27L2-VHrW17UviU:b8pum3s)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @stevendanna)



<!-- Sent from Reviewable.io -->

sherman-grewal commented:
bors r+
